### ISSUE TITLE
Doesn't seem to work with JavaScript source files

### ISSUE THREAD

**Issue by dschnare** [vytenisu/npm-dts#3]
Tried running this with a small test module and I get this:

```
> npm-dts generate -r . -e lib/index.js -L verbose

[npm-dts] [info] :  _________________________________________ 
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |          npm-dts v1.1.10                |
[npm-dts] [info] : |          by Vytenis Urbonavičius        |
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |_________________________________________|
[npm-dts] [info] :                                            
[npm-dts] [info] : Generating declarations for "/Users/{someone}/Workspace/sandbox/docs"...
[npm-dts] [verbose] : Locating OS Temporary Directory...
[npm-dts] [verbose] : OS Temporary Directory was located!
[npm-dts] [verbose] : Will now reset "tmp" directory...
[npm-dts] [verbose] : Cleaning up "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was cleaned!
[npm-dts] [verbose] : Preparing "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was prepared!
[npm-dts] [verbose] : Generating per-file typings using TSC...
[npm-dts] [error] : Generation of index.d.ts has failed!
[npm-dts] [verbose] : Cleaning up "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was cleaned!
[npm-dts] [verbose] : Deleting OS Temporary Directory...
[npm-dts] [verbose] : OS Temporary Directory was deleted!
[npm-dts] [error] : Generation failed!
```

My `tsconfig.json` is:

```
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true,
    "strict": true,
    "strictNullChecks": true,
    "outDir": "build"
  },
  "include": [
    "lib"
  ]
}
```

**Comment by vytenisu** [comment]
Thank you for report. I have not tried it with allowJs tbh. Will need to have a look.

**Comment by dschnare** [comment]
Interestingly TypeScript 3.7 has support for generating .d.ts files from JavaScript + JSDOC, however I've not been able to get it to work properly.

**Comment by dschnare** [comment]
Man, still can't get TypeScript 3.7 to generate .d.ts files. Such a pain. Anyway here's the debug output you requested:

```
> npm-dts generate -r . -e lib/index.js -L debug

[npm-dts] [info] :  _________________________________________ 
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |          npm-dts v1.1.10                |
[npm-dts] [info] : |          by Vytenis Urbonavičius        |
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |_________________________________________|
[npm-dts] [info] :                                            
[npm-dts] [info] : Generating declarations for "/Users/darren/sandbox/docs"...
[npm-dts] [verbose] : Locating OS Temporary Directory...
[npm-dts] [verbose] : OS Temporary Directory was located!
[npm-dts] [verbose] : Will now reset "tmp" directory...
[npm-dts] [verbose] : Cleaning up "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was cleaned!
[npm-dts] [verbose] : Preparing "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was prepared!
[npm-dts] [verbose] : Generating per-file typings using TSC...
[npm-dts] [debug] : tsc --declaration --emitDeclarationOnly --declarationDir "/var/folders/_w/p0438xw10_776hynr1tmrr340000gn/T/tmp-38454cg7InkRWir2a/npm-dts"
[npm-dts] [error] : Generation of index.d.ts has failed!
[npm-dts] [debug] : Error: 
lib/index.ts(1,9): error TS2451: Cannot redeclare block-scoped variable 'newUser'.
lib/index.ts(1,26): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i @types/node`.
lib/index.ts(3,1): error TS2304: Cannot find name 'exports'.
lib/index.ts(4,1): error TS2304: Cannot find name 'exports'.
lib/user.ts(13,10): error TS2451: Cannot redeclare block-scoped variable 'newUser'.
lib/user.ts(13,19): error TS7006: Parameter 'a' implicitly has an 'any' type.
lib/user.ts(21,1): error TS2304: Cannot find name 'exports'.
lib/user.ts(23,1): error TS2304: Cannot find name 'exports'.

[npm-dts] [verbose] : Cleaning up "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was cleaned!
[npm-dts] [verbose] : Deleting OS Temporary Directory...
[npm-dts] [verbose] : OS Temporary Directory was deleted!
[npm-dts] [error] : Generation failed!
```

**Comment by vytenisu** [comment]
Based on TS error in log above, you have TypeScript misconfiguration issue - not npm-dts.  The way npm-dts works is - it uses TSC for the heavy lifting and then joins up / alters a bit resulting declaration files to produce a single index.d.ts.

Could you share your tsconfig.json? Maybe I would spot something wrong in it.

**Comment by FDiskas** [comment]
would be nice to see some working examples

**Comment by dschnare** [comment]
My problem was that I didn't have `allowJs` enabled in the `tsconfig.json` file. However, now that TypeScript supports generating `.d.ts` files from JavaScript I've tried both your module and TypeScript directly. It works to an extent, but I can't get around types needing to be export types when the compiler hits an internal module that uses another internal module.

For example:
```
// a.js
class A {}
exports.A = A

// b.js
const { A } = require('./a')
class B {
  constructor () {
    this.a = new A()
  }
}

// index.js
exports.A = require('./a').A
exports.B = require('./b').B
```

When the compiler hits `b.js` in this example I get an error saying `A` is a private symbol and is not exported. I understand the error, just it's a real pain and seriously diminishes the utility of generating `.d.ts` files from JS.

My work around currently is using a command that uses the TypeScript compiler directly, but only one a single file at a time. I've bound it to the default build command in VSCode, so when I'm done with a JS file and I think it's ready for typings then I open the file and press `Cmd+Shift+B`. This generates the `.d.ts` file which is enough for me to start from.

**Comment by FDiskas** [comment]
@dschnare can you share that command line?

**Comment by vytenisu** [comment]
Finally got a chance to investigate the JS declaration problem as per last @dschnare comment. And managed to make it work by slightly changing JS code. For some reason TS gets angry if class name matches export name - it then thinks that I am trying to access a private non-exported thing.

Soooo... For now in order to avoid weird name clash issue - this is how I changed JS source files:

**index.js**
```
exports.XXX = require('./xxx').XXX
exports.YYY = require('./yyy').YYY
```

**xxx.js**
```
exports.XXX = class {}
```

**yyy.js**
```
const {XXX} = require('./xxx')

exports.YYY = class {
  constructor() {
    this.yyy = new XXX()
  }
}
```

**And I got this in DTS output:**

```
declare module 'npm-dts/test/sources/js/index' {
  export var XXX: {
      new (): {};
  };
  export var YYY: {
      new (): {
          yyy: {};
      };
  };

}
declare module 'npm-dts/test/sources/js/xxx' {
  export class XXX {
  }

}
declare module 'npm-dts/test/sources/js/yyy' {
  export class YYY {
      yyy: {};
  }
}

declare module 'npm-dts' {
  import main = require('npm-dts/index');
  export = main;
}
```

To be honest, I did not try consuming these typings but by the look of it - works like a charm. I am planning to release a new version of npm-dts very soon which will have e2e tests for covering this case. This will allow to see the code first hand.

**Comment by vytenisu** [comment]
@FDiskas, @dschnare, please find newly released 1.1.11 version: https://www.npmjs.com/package/npm-dts

Working JS example can be found here:
https://github.com/vytenisu/npm-dts/tree/master/test/sources/js

Tsconfig example:
https://github.com/vytenisu/npm-dts/blob/master/tsconfig.test.json

**Comment by FDiskas** [comment]
@vytenisu thanks but still getting errors
```
[npm-dts] [info] :  _________________________________________ 
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |          npm-dts v1.1.11                |
[npm-dts] [info] : |          by Vytenis Urbonavičius        |
[npm-dts] [info] : |                                         |
[npm-dts] [info] : |_________________________________________|
[npm-dts] [info] :                                            
[npm-dts] [info] : Generating declarations for "/Users/***/Projects/github/react-native-elements"...
[npm-dts] [verbose] : Locating OS Temporary Directory...
[npm-dts] [verbose] : OS Temporary Directory was located!
[npm-dts] [verbose] : Will now reset "tmp" directory...
[npm-dts] [verbose] : Cleaning up "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was cleaned!
[npm-dts] [verbose] : Preparing "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was prepared!
[npm-dts] [verbose] : Generating per-file typings using TSC...
[npm-dts] [error] : Generation of index.d.ts has failed!
[npm-dts] [verbose] : Cleaning up "tmp" directory...
[npm-dts] [verbose] : "tmp" directory was cleaned!
[npm-dts] [verbose] : Deleting OS Temporary Directory...
[npm-dts] [verbose] : OS Temporary Directory was deleted!
[npm-dts] [error] : Generation failed!
```
I'm trying to generate typings for project https://github.com/react-native-elements/react-native-elements

**Comment by vytenisu** [comment]
Will have a look when able. In the meantime, whenever debugging such issues, try running DTS generation using `-L debug`. This will output exact TSC command line used and error which prevented generation.

You would see something like in the above comment by @dschnare:

```
[npm-dts] [debug] : tsc --declaration --emitDeclarationOnly --declarationDir "/var/folders/_w/p0438xw10_776hynr1tmrr340000gn/T/tmp-38547lTt8mLH5dfJJ/npm-dts"
[npm-dts] [error] : Generation of index.d.ts has failed!
[npm-dts] [debug] : Error: 
error TS18003: No inputs were found in config file.......
```

**Comment by FDiskas** [comment]
Thanks for response - it worked. Typescript throwed some errors but typings still was generated.
Probably you script should at least try to solve/ignore errors :/ and see if there is some results are present

**Comment by vytenisu** [comment]
Good point and yes - that is doable!

You are right - npm-dts currently halts as soon as it encounters error output from TSC. I can introduce "force" mode which would continue for as long as there is d.ts produced despite any errors. Should be an easy win.

Can you paste errors that you get here by any chance?

**Comment by dschnare** [comment]
Wow good stuff. I'll investigate your workaround in a bit, but here's the command I've settled on that gets me part way there with the TypeScript compiler directly.

**package.json**
```json
{
  "name": "mypackage",
  "version": "0.0.1",
  "main": "./lib/index.js",
  "scripts": {
    "dts": "tsc --target es2018 --moduleResolution node --module commonjs --lib es2018,dom --allowJs --checkJs --declaration --emitDeclarationOnly"
  }
}
```

**.vscode/tasks.json**
```json
{
  // See https://go.microsoft.com/fwlink/?LinkId=733558
  // for the documentation about the tasks.json format
  "version": "2.0.0",
  "tasks": [
    {
      "type": "npm",
      "path": ".",
      "script": "dts -- --outDir ${fileDirname} ${file}",
      "problemMatcher": "$tsc",
      "group": {
        "kind": "build",
        "isDefault": true
      }
    }
  ]
}
```

Then in VSCode I just open the JS file I want to have TS declarations created for and press `Cmd+Shift+B`.

**Comment by dschnare** [comment]
Mind even with this command I still run into the "private symbols" issue from time to time, but as @vytenisu pointed out, this looks like it's a shortcoming of the TS compiler and how it parses JS to `.d.ts`. Oddly enough when I do get issues the `.d.ts` still gets generated albeit without any parameter or return type typings, but enough for a skeleton of the file.

**Comment by vytenisu** [comment]
I have just released **_npm-dts_ 1.2.0**.

It supports "-f" flag which would attempt generation despite errors. By using "-f" one needs to understand that output may not be 100% valid.

Let me know if this helps.
